src/legacy/ui/public/chrome/api/nav.ts

```diff
@@ -0,0 +1,163 @@
+
+export interface ChromeNavLinks {
+  untrackNavLinksForDeletedSavedObjects(deletedIds: string[]): void;
+  trackSubUrlForApp(linkId: string, parsedKibanaUrl: KibanaParsedUrl): void;
+}
+
+interface LegacyNavLinkProperties {
+  readonly subUrlBase: string;
+  readonly linkToLastSubUrl: boolean;
+}
+
+type LegacyNavLink = ChromeNavLink & LegacyNavLinkProperties;
+
+interface NavInternals {
+  nav: LegacyNavLink[];
+  appUrlStore: Storage;
+  trackPossibleSubUrl(url: string): void;
+}
+
+export function initChromeNavApi(chrome: any, internals: NavInternals) {
+  let coreNavLinks: ChromeStart['navLinks'];
+  onStart(({ core }) => (coreNavLinks = core.chrome.navLinks));
+
+  // These are legacy link properties that are not supported in the new platform.
+  // Index by link id for lookups when tracking sub urls.
+  const legacyNavProps = new Map(
+    internals.nav.map(
+      link =>
+        [link.id, pick(link, ['subUrlBase', 'linkToLastSubUrl'])] as [
+          string,
+          LegacyNavLinkProperties
+        ]
+    )
+  );
+
+  /**
+   * Clear last url for deleted saved objects to avoid loading pages with "Could not locate..."
+   */
+  chrome.untrackNavLinksForDeletedSavedObjects = (deletedIds: string[]) => {
+    function urlContainsDeletedId(url: string) {
+      const includedId = deletedIds.find(deletedId => {
+        return url.includes(deletedId);
+      });
+      if (includedId === undefined) {
+        return false;
+      }
+      return true;
+    }
+
+    coreNavLinks.getAll().forEach(link => {
+      const legacyProps = legacyNavProps.get(link.id)!;
+      if (legacyProps.linkToLastSubUrl && urlContainsDeletedId(link.url!)) {
+        setLastUrl(link, link.appUrl);
+      }
+    });
+  };
+
+  /**
+   * Manually sets the last url for the given app. The last url for a given app is updated automatically during
+   * normal page navigation, so this should only need to be called to insert a last url that was not actually
+   * navigated to. For instance, when saving an object and redirecting to another page, the last url of the app
+   * should be the saved instance, but because of the redirect to a different page (e.g. `Save and Add to Dashboard`
+   * on visualize tab), it won't be tracked automatically and will need to be inserted manually. See
+   * https://github.com/elastic/kibana/pull/11932 for more background on why this was added.
+   *
+   * @param id {String} - an id that represents the navigation link.
+   * @param kibanaParsedUrl {KibanaParsedUrl} the url to track
+   */
+  chrome.trackSubUrlForApp = (id: string, kibanaParsedUrl: KibanaParsedUrl) => {
+    const navLink = coreNavLinks.get(id);
+    if (navLink) {
+      setLastUrl(navLink, kibanaParsedUrl.getAbsoluteUrl());
+    }
+  };
+
+  internals.trackPossibleSubUrl = async function(url: string) {
+    const kibanaParsedUrl = absoluteToParsedUrl(url, chrome.getBasePath());
+
+    for (let link of coreNavLinks.getAll()) {
```

Comment:
I think I actually prefer using a mutable `link` variable here so that you can't introduce a bug by accidentally not using the most recent version of `link`.

On `forEach` vs. `for ... of` I'm a bit split. I've started leaning towards `for ... of` simply because you cannot use async/await with `forEach`.

Reply:
> On `forEach` vs. `for ... of` I'm a bit split. I've started leaning towards `for ... of` simply because you cannot use async/await with `forEach`.

Just for posterity, this is solved by using `map` instead of `forEach` with `Promise.all`. Use `map` when you want "parallel" async actions, and `for/of` when you want "serial" async actions:

```js
// "parallel"
await Promise.all(
  items.map(async item => {
    await something(item);
  });
);

// "serial"
for (const item of items) {
  await something(item);
}
```

Reply:
If you wanted this to be less imperative and avoid mutable variables, this could be written as:

```js
coreNavLinks.getAll().forEach(link => {
  const subUrlBase = legacyNavProps.get(link.id)!.subUrlBase;
  const active = kibanaParsedUrl.getAppRootPath().startsWith(subUrlBase);
  const updatedLink = coreNavLinks.update(link.id, { active })!;

  if (active) {
    setLastUrl(link, url);
    return;
  }

  const refreshedUrlLink = refreshLastUrl(link);
  const newGlobalState = kibanaParsedUrl.getGlobalState();

  if (newGlobalState) {
    injectNewGlobalState(link, kibanaParsedUrl.appId, newGlobalState);
  }
});
```

Not a requirement, just an option. :smile: